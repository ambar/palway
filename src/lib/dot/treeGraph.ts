import getPalIcon from '../getPalIcon'
import type {PalName} from '../palNames'
import {normalPalsByName} from '../pals'
import type {PalNode, PalNodePair} from '../search'
import unfold from '../unfold'

/**
 * Create a text graph of the matched nodes
 * @see https://graphviz.org/docs/graph/
 */
export const makeTreeGraph = (
  nodes: PalNode[],
  root: PalNode,
  parent: PalName,
  child: PalName,
  i18n: (key: string) => string,
) => {
  const results: string[] = []
  const visited = new Set<PalNode>()
  const attr = (node: PalNode) => {
    const palName = node.name
    const src = getPalIcon(normalPalsByName[palName].pal_dev_name)
    return [
      // 'fixedsize=true',
      `image="${src}"`,
      'labelloc=b',
      `label="${name(palName)}"`,
      palName === parent ? 'class=from' : '',
      // distinguish between root and descendant with child name
      node === root ? 'class=root' : '',
      node !== root && palName === child ? 'class=warn' : '',
    ]
      .filter(Boolean)
      .join(',')
  }
  const name = (palName: PalName) => {
    const pal = normalPalsByName[palName]
    return i18n(`palName_${pal.pal_dev_name}`)
  }
  let idCounter = 0
  const nextId = () => `i${idCounter++}`
  const definedSet = new Set<PalNode | PalNodePair>()
  const idMap = new Map<PalNode | PalNodePair, string>()
  // TODO: add id to tree
  const getID = (node: PalNode | PalNodePair) => {
    if (idMap.has(node)) {
      return idMap.get(node)
    }
    const id = nextId()
    idMap.set(node, id)
    return id
  }
  const defineNode = (node: PalNode) => {
    if (!node) {
      return
    }
    if (definedSet.has(node)) {
      return
    }
    const id = getID(node)
    definedSet.add(node)
    return `${id} [${attr(node)}]`
  }

  const definePair = (pair: PalNodePair) => {
    if (definedSet.has(pair)) {
      return
    }
    definedSet.add(pair)
    const l = getID(pair.left)
    const r = getID(pair.right)
    const pid = getID(pair)
    const cid = getID(pair.left.child)
    return [
      `${pid}[shape=circle,style=filled,label="",width=.05,height=.05]`,
      `${pid} -> ${cid}`,
      defineNode(pair.left.child),
      defineNode(pair.left),
      defineNode(pair.right),
      `${l} -> ${pid} [dir=none]`,
      `${r} -> ${pid} [dir=none]`,
    ]
      .filter(Boolean)
      .join('\n')
  }
  for (const node of nodes) {
    if (visited.has(node)) {
      continue
    }
    visited.add(node)
    const path = unfold(node, x => x.child)
    for (const node of path) {
      if (node.pair) {
        const r = definePair(node.pair)
        if (r) {
          results.push(r)
        }
      }
    }
  }
  return String.raw`
digraph Palway {
  node [fontname="Helvetica,Arial,sans-serif",style=dashed]
  edge [fontname="Helvetica,Arial,sans-serif",style=dashed,arrowsize=.9,arrowhead=empty]
  rankdir=BT
  node [shape=egg,width=1.5,height=1.5]
 ${results.join('\n')}
}`
}
