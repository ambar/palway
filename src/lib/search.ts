import {combosMapByChildName} from './combos'
import type {PalName} from './palNames'

export type PalNode = {
  name: PalName
  depth: number
  match?: boolean
  child?: PalNode
  pair?: PalNodePair
  mate?: PalNode
  parents: PalNodePair[]
}

export type PalNodePair = {
  [Symbol.iterator](): Generator<PalNode>
  left: PalNode
  right: PalNode
}

type LayerNodes = PalNodePair[]

export type PalTree = ReturnType<typeof createTree>

type CreateTreeVisitor = {
  create?(node: PalNode): void
  visitOnce?(node: PalNode): null | {
    /** @default false */
    skip?: boolean
    /** @default true */
    unvisited?: boolean
  }
}

/**
 * - Create a tree of all possible parents for a given child
 * - The tree is a directed graph, the root is the child
 */
export const createTree = (
  rootName: PalName,
  visitor: CreateTreeVisitor = {},
) => {
  // prevent infinite loops
  const visited = new Set<PalName>()
  const makeNode = (name: PalName, depth: number) => {
    const node: PalNode = {name, depth, parents: []}
    visitor.create?.(node)
    return node
  }
  const makeParents = (node: PalNode, depth: number): PalNodePair[] => {
    return (combosMapByChildName.get(node.name) ?? []).map(([a, b]) => {
      const left = makeNode(a, depth)
      const right = makeNode(b, depth)
      left.child = node
      right.child = node
      left.mate = right
      right.mate = left
      const pair = {
        left,
        right,
        *[Symbol.iterator]() {
          yield left
          yield right
        },
      }
      left.pair = pair
      right.pair = pair
      return pair
    })
  }
  const layers: LayerNodes[] = []
  const root = makeNode(rootName, 0)
  while (true) {
    const layer = layers[layers.length - 1]
    const neighbors: PalNodePair[] = []
    const visit = (node: PalNode) => {
      if (!visited.has(node.name)) {
        const r = visitor.visitOnce?.(node) || {}
        if (!r.unvisited) {
          visited.add(node.name)
        }
        if (r.skip || node.match) {
          return
        }
        node.parents = makeParents(node, layers.length + 1)
        neighbors.push(...node.parents)
      }
    }
    if (!layer) {
      visit(root)
    } else {
      for (const pair of layer) {
        for (const node of pair) {
          visit(node)
        }
      }
    }
    if (!neighbors.length) {
      break
    }
    layers.push(neighbors)
  }
  return {root, layers}
}

type Visitor = {
  node?: (node: PalNode) => void
  pair?: (pair: PalNodePair) => void
  layer?: (layer: PalNodePair[]) => void
}

export const walk = (tree: PalTree, callback: Visitor) => {
  callback.node?.(tree.root)
  for (const layer of tree.layers) {
    callback.layer?.(layer)
    for (const pair of layer) {
      callback.pair?.(pair)
      for (const node of pair) {
        callback.node?.(node)
      }
    }
  }
}
