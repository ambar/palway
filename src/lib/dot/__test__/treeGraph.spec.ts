import {expect, test} from 'vitest'
import {createTree} from '../../search'
import {makeTreeGraph} from '../treeGraph'

const makeTreeResult = (root, parent) => {
  const nodes = []
  const tree = createTree(root, {
    create(node) {
      if (node.name === parent) {
        node.match = true
        nodes.push(node)
      }
    },
    visitOnce(node) {
      if (node.mate?.name === root) {
        return {break: true, unvisited: true}
      }
    },
  })
  return {tree, nodes}
}

test('makeTreeGraph', () => {
  const parent = 'Jetragon'
  const child = 'Lamball'
  const {tree, nodes} = makeTreeResult(child, parent)
  expect(
    makeTreeGraph(nodes.slice(0, 1), tree.root, parent, child, v => v),
  ).toMatchSnapshot()
})
