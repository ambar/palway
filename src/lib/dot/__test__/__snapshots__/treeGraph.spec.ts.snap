// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`makeTreeGraph 1`] = `
"
digraph Palway {
  node [fontname="Helvetica,Arial,sans-serif",style=dashed]
  edge [fontname="Helvetica,Arial,sans-serif",style=dashed,arrowsize=.9,arrowhead=empty]
  rankdir=BT
  node [shape=egg,height=1.5,width=1.5,labelloc=b];
 i2[shape=circle,style=filled,label="",width=.05,height=.05]
i2 -> i3
i3 [image="/src/assets/images/pal_icons_200w/T_FlowerDoll_icon_normal.png",labelloc=b,label="palName_FlowerDoll"]
i0 [image="/src/assets/images/pal_icons_200w/T_PinkCat_icon_normal.png",labelloc=b,label="palName_PinkCat"]
i1 [image="/src/assets/images/pal_icons_200w/T_JetDragon_icon_normal.png",labelloc=b,label="palName_JetDragon",class=from]
i0 -> i2 [dir=none]
i1 -> i2 [dir=none]
i5[shape=circle,style=filled,label="",width=.05,height=.05]
i5 -> i6
i6 [image="/src/assets/images/pal_icons_200w/T_Boar_icon_normal.png",labelloc=b,label="palName_Boar"]
i4 [image="/src/assets/images/pal_icons_200w/T_SheepBall_icon_normal.png",labelloc=b,label="palName_SheepBall",class=warn]
i4 -> i5 [dir=none]
i3 -> i5 [dir=none]
i8[shape=circle,style=filled,label="",width=.05,height=.05]
i8 -> i9
i9 [image="/src/assets/images/pal_icons_200w/T_MopBaby_icon_normal.png",labelloc=b,label="palName_MopBaby"]
i7 [image="/src/assets/images/pal_icons_200w/T_SheepBall_icon_normal.png",labelloc=b,label="palName_SheepBall",class=warn]
i7 -> i8 [dir=none]
i6 -> i8 [dir=none]
i11[shape=circle,style=filled,label="",width=.05,height=.05]
i11 -> i12
i12 [image="/src/assets/images/pal_icons_200w/T_WindChimes_Ice_icon_normal.png",labelloc=b,label="palName_WindChimes_Ice"]
i10 [image="/src/assets/images/pal_icons_200w/T_WindChimes_icon_normal.png",labelloc=b,label="palName_WindChimes"]
i10 -> i11 [dir=none]
i9 -> i11 [dir=none]
i14[shape=circle,style=filled,label="",width=.05,height=.05]
i14 -> i15
i15 [image="/src/assets/images/pal_icons_200w/T_PinkCat_icon_normal.png",labelloc=b,label="palName_PinkCat"]
i13 [image="/src/assets/images/pal_icons_200w/T_ChickenPal_icon_normal.png",labelloc=b,label="palName_ChickenPal"]
i13 -> i14 [dir=none]
i12 -> i14 [dir=none]
i17[shape=circle,style=filled,label="",width=.05,height=.05]
i17 -> i18
i18 [image="/src/assets/images/pal_icons_200w/T_SheepBall_icon_normal.png",labelloc=b,label="palName_SheepBall",class=root]
i16 [image="/src/assets/images/pal_icons_200w/T_SheepBall_icon_normal.png",labelloc=b,label="palName_SheepBall",class=warn]
i15 -> i17 [dir=none]
i16 -> i17 [dir=none]
}"
`;
